src/register-service-worker.js

```diff
@@ -0,0 +1,162 @@
+/* eslint-env browser */
+
+(function() {
+  var constants = require('./lib/constants.js');
+
+  var swScript = document.currentScript.dataset.serviceWorker;
+  var manifestAttribute = document.documentElement.getAttribute('manifest');
+
+  if (swScript && manifestAttribute && 'serviceWorker' in navigator) {
+    var manifestUrl = (new URL(manifestAttribute, location.href)).href;
+
+    var idb = require('idb');
+
+    idb.open(constants.DB_NAME, constants.DB_VERSION, function(upgradeDB) {
+      if (upgradeDB.oldVersion === 0) {
+        Object.keys(constants.OBJECT_STORES).forEach(function(objectStore) {
+          upgradeDB.createObjectStore(constants.OBJECT_STORES[objectStore]);
+        });
+      }
+    }).then(function(db) {
+      return Promise.all([
+        updateManifestAssociationForCurrentPage(db, manifestUrl),
+        handlePossibleManifestUpdate(db, manifestUrl)
+      ]);
+    }).then(function() {
+      return navigator.serviceWorker.register(swScript);
+    });
+  }
+
+  /**
+   * Caches the Responses for one or more URLs, using the Cache Storage API.
+   *
+   * @param {Array.<String>} urls
+   * @returns {Promise.<T>}
+   */
+  function addToCache(urls) {
+    return caches.open(constants.CACHE_NAME).then(function(cache) {
+      return cache.addAll(urls.map(function(url) {
+        // These caching requests need to always result in a fetch() even if
+        // there's a service worker that might otherwise treat them differently.
+        // Let's use this special request header to let the service worker know.
+        return new Request(url, {headers: {'X-Use-Network': true}});
```

Comment:
Check out 18.4 https://html.spec.whatwg.org/multipage/browsers.html#downloading-or-updating-an-application-cache

Appcache resources fetches are different to the above. Also be aware that some of the fetch spec defaults are different to the `new Request()` defaults. For instance, mode is "no-cors" by default, and since no other value is specified in the appcache spec, I guess fetches are all "no-cors".